smart-contracts/ethereum/contracts/MarketPlace.sol

```diff
@@ -1,43 +1,105 @@
     string name;
-    appVersion[] versions;
+    string category;
+    mapping  (bytes32 => AppVersion)  versions;
+    bytes32[] versionList;
   }
   
   mapping (uint => App) apps;
-  mapping (address => bool) operators;
+  uint[] app_ids;
+  mapping (address => bool) delegates;
+
+  address republicContract;
+  
+  event AppSubmitted(uint id);
+  event VersionSubmitted(uint app_id, bytes32 version);
+  event AppVoting(uint app_id, bytes32 version, int8 vote);
+  
+  function MarketPlace() public {
+    delegates[msg.sender] = true; // only one delegate for now until republic contract is fully functional
+  }
+
+  function setRepublicContract (address _contract_address) public onlyDelegates returns(bool res) {
+    republicContract = _contract_address;
+
+    return true;
+  }
+  
+  function refreshDelegatesList() public pure returns(bool res) { // this should be called after delegates elections
+    return true;
+  }
   
-  function Marketplace() public {
-    
+
+  function submitAppForReview(string _name, bytes32 _version, string _category, string _files, string _checksum, uint _permissions) public returns(bool res) {
+    uint new_id = app_ids.length;
+    app_ids.push(new_id);
+    bytes32[] memory versionList;
+
+    apps[new_id] = App(new_id, msg.sender, _name, _category, versionList);
+    AppSubmitted(new_id);
+    submitVersionForReview(new_id, _version, _files, _checksum, _permissions);
+
+    return true;
   }
 
-  function submitAppForReview() public pure returns(bool res) { // TODO: string name, uint version, string files, string checksum
+  function submitVersionForReview(uint _id, bytes32 _version, string _files, string _checksum, uint _permissions) public returns(bool res) {
+    require(apps[_id].id == _id);
+    require(apps[_id].versions[_version].version != _version);
+    address[] memory voters;
+
+    apps[_id].versions[_version] = AppVersion(_version, _files, voters, _checksum, _permissions);
+    apps[_id].versionList.push(_version);
+    VersionSubmitted(_id, _version);
+
     return true;
   }
+  
+
 
-  function vote() onlyOperator public view returns(bool res) { // TODO: uint id, bool vote
+  function voteForApp(uint _id, bytes32 _version, int8 _vote) public onlyDelegates returns(bool res) {
+    require (apps[_id].id == _id && apps[_id].versions[_version].version == _version);
+    require (_vote == -1 || _vote == 1);
+
+    if (apps[_id].versions[_version].votes[msg.sender] == 0) {
+      apps[_id].versions[_version].voters.push(msg.sender);
+    }
+
+    apps[_id].versions[_version].votes[msg.sender] = _vote;
+
+    AppVoting(_id, _version, _vote);
     return true;
   }
-}
+  
+  function getApp(uint _id, bytes32 _version) public view returns(address owner, string name, string category, string files, int votes) {
+    require (apps[_id].id == _id && apps[_id].versions[_version].version == _version);
+
+    int8 vote_sum = 0;
+
+    for(uint i; i < apps[_id].versions[_version].voters.length; i++) {
```

Comment:
Feel like this should be calculated on write not on read, what do you think?

Reply:
Also prob want to assign `apps[_id].versions[_version]` to a var for clarity (optimization too?)

Reply:
1- Right.. i thought of adding it on write instead of read. but thought that it might lead to a race condition since vote_sum is going to be a shared variable and if two delegates voted at the same time then vote_sum might end up with a wrong value.
2- Yes i will assign it to a variable. tho i don't think it's an optimization since accessing  apps[_id].versions[_version] is O(1) and writing it to a variable might be more costly

Reply:
1 - Is that a thing? We cant expect transactions to be procedural and calculate properly?
2 - Makes sense